coresimd/x86/cpuid.rs

```diff
@@ -86,26 +86,48 @@ pub fn has_cpuid() -> bool {
     }
     #[cfg(target_arch = "x86")]
     {
-        use coresimd::x86::{__readeflags, __writeeflags};
-
-        // On `x86` the `cpuid` instruction is not always available.
-        // This follows the approach indicated in:
-        // http://wiki.osdev.org/CPUID#Checking_CPUID_availability
         unsafe {
-            // Read EFLAGS:
-            let eflags: u32 = __readeflags();
-
-            // Invert the ID bit in EFLAGS:
-            let eflags_mod: u32 = eflags | 0x0020_0000;
-
-            // Store the modified EFLAGS (ID bit may or may not be inverted)
-            __writeeflags(eflags_mod);
-
-            // Read EFLAGS again:
-            let eflags_after: u32 = __readeflags();
-
-            // Check if the ID bit changed:
-            eflags_after != eflags
+            // On `x86` the `cpuid` instruction is not always available.
+            // This follows the approach indicated in:
+            // http://wiki.osdev.org/CPUID#Checking_CPUID_availability
+            // https://software.intel.com/en-us/articles/using-cpuid-to-detect-the-presence-of-sse-41-and-sse-42-instruction-sets/
+            // which detects whether `cpuid` is available by checking whether the 21st bit of the EFLAGS register is modifiable or not.
+            // If it is, then `cpuid` is available.
+            let eax: i32;
+            asm!(r#"
+                # Save a copy of the original eflags that we will restore later:
+                pushfd
+                # Copy eflags to ecx and eax:
+                pushfd
+                pop %eax
+                mov %ecx, %eax
+                # Flip 21st bit and write back to eflags register:
+                xor %eax, 0x200000
+                push %eax
+                popfd
+                # Read eflags register again:
+                pushfd
+                pop %eax
+                # If cpuid is available, the bit will still be flipped
+                # and it will be the only bit modified.
+                #
+                # xor with the original eflags should produce a 1 in
+                # the 21st bit in this case, and zeros for all other bits:
+                xor %eax, %ecx
+                # So if the value of the 21st bit is 1, cpuid is available,
+                # and if it is zero, it isn't because we didn't manage to
+                # modify it:
+                shrd %eax, 21
```

Comment:
So the problem is that this is the instruction for doubles, it should be shr, or shrl, but none of these appear to work.